runtime/devnet/src/extensions/mod.rs

```diff
@@ -0,0 +1,1160 @@
+		52 => {
+			let key = versioned_construct_assets_key(version, call_index, params)?;
+			Ok(RuntimeStateKeys::Assets(key))
+		},
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn versioned_construct_assets_call(
+	version: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<pallet_assets::Call<Runtime, TrustBackedAssetsInstance>, DispatchError> {
+	match version {
+		V0 => v0::assets::construct_assets_call(call_index, params),
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn versioned_construct_assets_key(
+	version: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<AssetsKeys, DispatchError> {
+	match version {
+		V0 => v0::assets::construct_assets_key(call_index, params),
+		_ => Err(DispatchError::Other("UnknownFunctionId")),
+	}
+}
+
+fn read_state<T, E>(
+	env: &mut Environment<E, BufInBufOutState>,
+	version: u8,
+	pallet_index: u8,
+	call_index: u8,
+	params: Vec<u8>,
+) -> Result<(), DispatchError>
+where
+	T: pallet_contracts::Config
+		+ pallet_assets::Config<TrustBackedAssetsInstance, AssetId = AssetId>
+		+ pallet_nfts::Config<CollectionId = CollectionId, ItemId = ItemId>
+		+ cumulus_pallet_parachain_system::Config
+		+ frame_system::Config<AccountId = sp_runtime::AccountId32>,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " read_state |";
+	let key = construct_key(version, pallet_index, call_index, params)?;
+	let result = match key {
+		RuntimeStateKeys::Nfts(key) => read_nfts_state::<T, E>(key, env),
+		RuntimeStateKeys::ParachainSystem(key) => read_parachain_system_state::<T, E>(key, env),
+		RuntimeStateKeys::Assets(key) => read_assets_state::<T, E>(key, env),
+	}?
+	.encode();
+	log::trace!(
+		target:LOG_TARGET,
+		"{} result: {:?}.", LOG_PREFIX, result
+	);
+	env.write(&result, false, None)
+}
+
+fn send_xcm<T, E>(env: &mut Environment<E, BufInBufOutState>) -> Result<(), DispatchError>
+where
+	T: pallet_contracts::Config
+		+ frame_system::Config<
+			RuntimeOrigin = RuntimeOrigin,
+			AccountId = AccountId,
+			RuntimeCall = RuntimeCall,
+		>,
+	E: Ext<T = T>,
+{
+	const LOG_PREFIX: &str = " send_xcm |";
+	// Read the input as CrossChainMessage.
+	let xc_call: CrossChainMessage = env.read_as::<CrossChainMessage>()?;
+	// Determine the call to dispatch.
+	let (dest, message) = match xc_call {
+		CrossChainMessage::Relay(message) => {
+			let dest = Location::parent().into_versioned();
+			let assets: Asset = (Here, 10 * UNIT).into();
+			let beneficiary: Location =
+				AccountId32 { id: (env.ext().address().clone()).into(), network: None }.into();
+			let message = Xcm::builder()
+				.withdraw_asset(assets.clone().into())
+				.buy_execution(assets.clone(), Unlimited)
+				.transact(
+					SovereignAccount,
+					Weight::from_parts(250_000_000, 10_000),
+					message.encode().into(),
+				)
+				.refund_surplus()
+				.deposit_asset(assets.into(), beneficiary)
+				.build();
+			(dest, message)
+		},
+	};
+	// TODO: revisit to replace with signed contract origin
```

Comment:
Pretty sure this was only required for Rococo and was fixed.

Duplicate chain extensions across runtimes will become a problem, so we should try to tackle sooner rather than later.

Reply:
Created an issue for both, revisit to replace with signed contract origin (#106) and resolving the duplicate chain extension (#107)